agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1411 @@
+            match instance_info.connectivity_status {
+                InstanceConnectivityStatus::Online => {}
+                InstanceConnectivityStatus::Offline(instant) => {
+                    let time_offline = instant.elapsed().as_secs();
+                    // If instance has been offline for longer than the grace period or it is unshared, terminate the associated device plugin
+                    // TODO: make grace period configurable
+                    if time_offline >= SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS {
+                        trace!("check_offline_status - instance {} has been offline too long ... terminating device plugin", instance);
+                        device_plugin_service::terminate_device_plugin_service(
+                            &instance,
+                            self.instance_map.clone(),
+                        )
+                        .await
+                        .unwrap();
+                        k8s::try_delete_instance_arc(
+                            kube_interface_clone.clone(),
+                            instance.clone(),
+                            self.config.metadata.namespace.as_ref().unwrap().clone(),
+                        )
+                        .await?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Takes in a list of discovered devices and determines if there are any new devices or no longer visible devices.
+    /// For each new device, it creates a DevicePluginService.
+    /// For each previously visible device that was no longer discovered, it calls a function that updates the InstanceConnectivityStatus
+    /// of the instance or deletes it if it is a local device.
+    pub async fn handle_discovery_results(
+        &self,
+        kube_interface: Arc<Box<dyn k8s::KubeInterface>>,
+        discovery_results: Vec<Device>,
+        is_local: bool,
+        device_plugin_builder: Box<dyn DevicePluginBuilderInterface>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        trace!(
+            "handle_discovery_results - for config {} with discovery results {:?}",
+            self.config.metadata.name,
+            discovery_results
+        );
+        let currently_visible_instances: HashMap<String, Device> = discovery_results
+            .iter()
+            .map(|discovery_result| {
+                let id = generate_instance_digest(&discovery_result.id, !is_local);
+                let instance_name = get_device_instance_name(&id, &self.config.metadata.name);
+                (instance_name, discovery_result.clone())
+            })
+            .collect();
+        INSTANCE_COUNT_METRIC
+            .with_label_values(&[&self.config.metadata.name, &is_local.to_string()])
+            .set(currently_visible_instances.len() as i64);
+        // Update the connectivity status of instances and return list of visible instances that don't have Instance CRs
```

Comment:
nit: Instance CRDs

Reply:
I think of Instances, like an instance for /dev/video1 as an Instance Custom Resource, while Akri has one Instance Custom Resource Definition or blueprint for Instance CRs

Reply:
👍